Network: add certificate to peer diagnostics

gerardsn opened the pull request:
this adds the peer's certificate in PEM format to
```
GET /internal/network/v1/diagnostics/peers
```
This provides us with the certificate when it needs to be added to a denylist.
I have added a `PeerKey` to prevent masking the certificate by `PeerID` spoofing.



```
   "17a9f6b8-f4fb-4beb-86d7-9c4eafac8e9a()@nuts-stable.reinkrul.nl:8443": {
      ....
      same as below, only difference is missing nodeDID in identifier (bootstrap node in this case)
      ....
    },
    "17a9f6b8-f4fb-4beb-86d7-9c4eafac8e9a(did:nuts:7eaDstczhnigW9uW1JRzYF2u5KhTntFTnee8vGDHokD)@nuts-stable.reinkrul.nl:8443": {
        "certificate": "-----BEGIN CERTIFICATE-----\nMIIBjzCCATagAwIBAgIJAP4dHF14qi40MAoGCCqGSM49BAMCMDIxMDAuBgNVBAMM\nJ051dHMgU3RhYmxlIERldmVsb3BtZW50IE5ldHdvcmsgUm9vdCBDQTAeFw0yMzA2\nMjExNDU3MzlaFw0yNDA2MjAxNDU3MzlaMCIxIDAeBgNVBAMMF251dHMtc3RhYmxl\nLnJlaW5rcnVsLm5sMFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEqhKmwi+eX2gc\nnGFHMr66R7kxMp/NhXpYk5aKoYLegjkT1KNYcPNjsiKz8ZjnahSg3pgROyOKwr/2\nmNKCePaJkqNFMEMwHQYDVR0lBBYwFAYIKwYBBQUHAwEGCCsGAQUFBwMCMCIGA1Ud\nEQQbMBmCF251dHMtc3RhYmxlLnJlaW5rcnVsLm5sMAoGCCqGSM49BAMCA0cAMEQC\nIEBLO2+rjdBsFsOw2KyVIBDfSMm43QL/egTaelrurWiIAiACsRj50OD92D0GR5FZ\nmVDOXyZBbXNpV5r19iEpndV3AA==\n-----END CERTIFICATE-----\n",
        "peers": [
            "9471bad9-6375-4c81-866f-92acb142e5aa",
            "746ff0ce-20f8-4228-ba45-51bc9f5c5a6b",
            "523bcabe-ac38-425d-a127-81e017416cbc",
            "2a9b2b69-0241-4dfe-8e78-e2c339cb5e32",
            "f5abe675-b983-4d46-9eb4-d41ee9cb3ff0",
            "f5abe675-b983-4d46-9eb4-d41ee9cb3ff0",
            "b78668cd-3846-45df-bb57-10b7421a366b",
            "64ef53c9-28cf-4e13-928a-5a90849ff3c9",
            "10058c59-b338-4c90-966b-923521df148c",
            "76d04f8a-fd1a-486d-bb61-d4b71176ccb2",
            "1f5ef3b3-0a4f-45d2-be98-26c2b6347a9b",
            "a0bf9065-f63e-4354-a19a-252128d2b694",
            "321830ca-0605-46b5-a9e9-4a3aa8ba9566-bootstrap",
            "321830ca-0605-46b5-a9e9-4a3aa8ba9566"
        ],
        "softwareID": "https://github.com/nuts-foundation/nuts-node",
        "softwareVersion": "master (441cdda4596062d3d4b4a297dac16bc174f2c0e3)",
        "transactionNum": 27685,
        "uptime": 5670
    },
```

todo:
- [x] tests

reinkrul commented on docs/_static/network/v1.yaml:
```diff
@@ -292,6 +292,10 @@ components:
           type: string
+        certificate:
+          description: >
+            Certificate presented during TLS handshake to setup the existing connection.
```
This sounds convoluted (learned the word from @beardedfoo).

Just: the peer's certificate?

reinkrul commented on network/transport/types.go:
```diff
@@ -41,6 +43,14 @@ func (p PeerID) String() string {
+type PeerKey string
+
+// String returns the PeerKey as string.
+func (p PeerKey) String() string {
```
not a very useful type?

gerardsn replied:
The changes to diagnostics manager resulted in some return values changing from `map[PeerID]any` to `map[string]any`. I didn't likt the ambiguity there so added the `PeerKey`

reinkrul commented on network/transport/types.go:
```diff
@@ -94,6 +116,8 @@ type Diagnostics struct {
 	// Proprietary implementations could specify the product's or vendor's name.
 	SoftwareID string `json:"softwareID"`
+	// Certificate contains the certificate presented by the peer during the TLS handshake.
+	Certificate string `json:"certificate"`
```
could've been OK to just return it as `[]byte` (meaning DER, no need to PEM encode?)

But PEM is probably more recognizable, so OK.

gerardsn replied:
PEM is readable when printed, makes my life a bit easier

reinkrul commented on network/transport/v2/peer_diagnostics.go:
```diff
@@ -57,25 +57,27 @@ func (m *peerDiagnosticsManager) start(ctx context.Context, broadcastInterval ti
 		NumberOfTransactions: received.NumberOfTransactions,
 		SoftwareVersion:      received.SoftwareVersion,
 		SoftwareID:           received.SoftwareID,
+		Certificate:          peer.CertificateAsPem(),
```
```suggestion
		Certificate:          peer.CertificateAsPem(), // enriches received diagnostics (not sent by peer)
```

woutslakhorst reviewed (changes requested):
This is a braking change, if you want to show address and DID as well, add them as additional fields next to the certificate.

woutslakhorst commented:
Diagnostics are used when investigating so they should not have any security considerations. I would prefer the additional fields for v5. v6 could change things but that is of a later concern.

gerardsn commented:
since this will be the only place showing the certificate it is no longer strictly diagnostics. Certificate information could then be hidden by PeerID spoofing, what I was trying to prevent. 

`/status/diagnostics` falls into the same category, but maybe the address book is a better place for the cert? This then always shows the certificate used on the last successful connection? I'd have to think a bit longer about the security considerations.

woutslakhorst commented:
The underlying problem is the identification of a peer and how we present it. Different places use different representations.

What if we:

- use the DID as peerID for authenticated connections
- use the certificate as peerID for secured connections without DID
- use peerID for unsecured/unknown connections

gerardsn commented:
Reverted the PeerKey addition and added `peer.Address` and `Peer.NodeDID` to the diagnostics.
peer diagnostics currently look like:
```
    "43e92fc4-b81b-459b-bba3-6f6bfc40b201": {
        "address": "nuts-stable.reinkrul.nl:8443",
        "certificate": "-----BEGIN CERTIFICATE-----\nMIIBjzCCATagAwIBAgIJAP4dHF14qi40MAoGCCqGSM49BAMCMDIxMDAuBgNVBAMM\nJ051dHMgU3RhYmxlIERldmVsb3BtZW50IE5ldHdvcmsgUm9vdCBDQTAeFw0yMzA2\nMjExNDU3MzlaFw0yNDA2MjAxNDU3MzlaMCIxIDAeBgNVBAMMF251dHMtc3RhYmxl\nLnJlaW5rcnVsLm5sMFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEqhKmwi+eX2gc\nnGFHMr66R7kxMp/NhXpYk5aKoYLegjkT1KNYcPNjsiKz8ZjnahSg3pgROyOKwr/2\nmNKCePaJkqNFMEMwHQYDVR0lBBYwFAYIKwYBBQUHAwEGCCsGAQUFBwMCMCIGA1Ud\nEQQbMBmCF251dHMtc3RhYmxlLnJlaW5rcnVsLm5sMAoGCCqGSM49BAMCA0cAMEQC\nIEBLO2+rjdBsFsOw2KyVIBDfSMm43QL/egTaelrurWiIAiACsRj50OD92D0GR5FZ\nmVDOXyZBbXNpV5r19iEpndV3AA==\n-----END CERTIFICATE-----\n",
        "nodeDID": "did:nuts:7eaDstczhnigW9uW1JRzYF2u5KhTntFTnee8vGDHokD",
        "peers": [
            "746ff0ce-20f8-4228-ba45-51bc9f5c5a6b",
            "b35ee900-468b-4c65-a700-d0ca9fec43ad",
            "37f1e471-8fb1-4102-b4d7-c4b29abf97b8",
            "2a9b2b69-0241-4dfe-8e78-e2c339cb5e32",
            "f2561e2a-6df6-4f4e-9e0a-79546d63e962",
            "b35ee900-468b-4c65-a700-d0ca9fec43ad",
            "b78668cd-3846-45df-bb57-10b7421a366b",
            "10058c59-b338-4c90-966b-923521df148c",
            "9471bad9-6375-4c81-866f-92acb142e5aa",
            "b00e502e-ff9c-48d2-8b70-38a79d2d8dcd",
            "523bcabe-ac38-425d-a127-81e017416cbc",
            "84964db0-16be-464f-83c5-2629d7c9b4f3",
            "460a4614-1752-47e3-b395-2dd365a2c5e3-bootstrap",
            "460a4614-1752-47e3-b395-2dd365a2c5e3"
        ],
        "softwareID": "https://github.com/nuts-foundation/nuts-node",
        "softwareVersion": "master (45a4b9d99c5abcb7e0379fdbb646764d2ebb97e2)",
        "transactionNum": 27866,
        "uptime": 6485
    },
```
